proxy/grpcproxy/lease.go

```diff
@@ -46,42 +94,209 @@ func (lp *leaseProxy) LeaseTimeToLive(ctx context.Context, rr *pb.LeaseTimeToLiv
-			if err != nil {
-				return
-			}
-			err = lc.Send(rr)
-			if err != nil {
-				return
-			}
+		defer func() { stopc <- struct{}{} }()
+		errCh <- lps.recvLoop()
+	}()
+
+	// watches all the listeningFromServer() go routines
+	go func() {
+		defer func() { stopc <- struct{}{} }()
+		lps.sendLoop()
+	}()
+
+	// tears down LeaseKeepAlive stream if leader goes down or entire leaseProxy is terminated.
+	go func() {
+		defer func() { stopc <- struct{}{} }()
+		select {
+		case <-leaderc:
+		case <-cctx.Done():
+		case <-lp.ctx.Done():
 		}
 	}()
 
-	for {
-		rr, err := lc.Recv()
-		if err != nil {
-			return err
+	<-stopc
+	ccancel()
+
+	go func() {
+		<-stopc
+		<-stopc
+		lps.close()
+		lp.wg.Done()
+	}()
+
+	select {
+	case <-leaderc:
+		return rpctypes.ErrNoLeader
+	case err = <-errCh:
+		return err
+	default:
+		return lp.ctx.Err()
+	}
+}
+
+type leaseProxyStream struct {
+	stream pb.Lease_LeaseKeepAliveServer
+	client *clientv3.Client
+
+	wg sync.WaitGroup
+	// mu protects keepAliveLeases
+	mu sync.RWMutex
+	// keepAliveLeases map keeps track of on-going lease renewal IDs and corresponding chans.
+	// the chan tracks number of keepalive requests received for a given lease ID.
+	// it is used to control number of keepalive responses to be sent back to the client
+	keepAliveLeases map[int64]chan struct{}
+	// lrpCh is a channel to receive lease keepalive response from etcd backend
+	lrpCh chan *pb.LeaseKeepAliveResponse
+
+	ctx context.Context
+}
+
+func (lps *leaseProxyStream) recvLoop() error {
+	defer func() {
+		fmt.Printf("leaseProxyStream %v closed \n", lps.stream)
+	}()
+	for lps.ctx.Err() == nil {
+		fmt.Printf("leaseProxyStream %v receiving msg...\n", lps.stream)
+		rr, err := lps.stream.Recv()
+		fmt.Printf("leaseProxyStream %v received msg %v \n", lps.stream, rr)
+		if err == io.EOF {
+			return nil
 		}
-		err = stream.Send(rr)
 		if err != nil {
 			return err
 		}
+		lps.serveKeepAliveRequest(rr)
+	}
+	return lps.ctx.Err()
+}
+
+func (lps *leaseProxyStream) serveKeepAliveRequest(rr *pb.LeaseKeepAliveRequest) {
+	fmt.Printf("leaseProxyStream.serveKeepAliveRequest() req %v ...\n", rr)
+	lps.mu.Lock()
+	defer lps.mu.Unlock()
+	msgsCh, ok := lps.keepAliveLeases[rr.ID]
+	if !ok {
+		msgsCh = make(chan struct{}, 16)
+		lps.keepAliveLeases[rr.ID] = msgsCh
+		lps.wg.Add(1)
+		go func() {
+			defer lps.wg.Done()
+			lps.listeningFromServer(rr.ID)
+		}()
+	}
+	select {
+	case msgsCh <- struct{}{}:
+		fmt.Printf("leaseProxyStream.serveKeepAliveRequest() put items in msgsCh \n")
+	default:
+	}
+	fmt.Printf("leaseProxyStream.serveKeepAliveRequest() req %v done \n", rr)
+}
+
+func (lps *leaseProxyStream) listeningFromServer(leaseID int64) {
+	fmt.Printf("leaseProxyStream.listeningFromServer() %v ...\n", leaseID)
+	var (
+		msgsCh <-chan struct{}
+		ticker <-chan time.Time
+	)
+	cctx, ccancel := context.WithCancel(lps.ctx)
+	defer func() {
+		// clean up
+		ccancel()
+		lps.mu.Lock()
+		delete(lps.keepAliveLeases, leaseID)
```

Comment:
Suppose that I get rid of the ticker. the keepAliveLoop() only exits if context is cancelled or `lps.client.KeepAlive()` closes the response channel. In this case, a new ka request finds an entry in lps.keepAliveLeases and passes through right before the `keepAliveLoop()` defer function deletes the entry in lps.keepAliveLeases. in this case, what should we deal with the new request that's not being processed? should the keepAliveLoop() clean up logic check if there are outstanding keepalive requests and send some kinda error back to stream?